File: app/src/main/java/org/metabrainz/mobile/data/sources/api/entities/mbentity/ReleaseGroup.java
package org.metabrainz.mobile.data.sources.api.entities.mbentity;

import com.google.gson.annotations.SerializedName;

import org.metabrainz.mobile.data.sources.api.entities.ArtistCredit;
import org.metabrainz.mobile.data.sources.api.entities.Link;

import java.util.ArrayList;
import java.util.List;

public class ReleaseGroup extends MBEntity {
    private String title;
    @SerializedName("primary-type")
    private String primaryType;
    @SerializedName("secondary-types")
    private final List<String> secondaryTypes = new ArrayList<>();
    private int count;
    //TODO: Implement correct wrapper JSON
    @SerializedName("artist-credit")
    private final List<ArtistCredit> artistCredits = new ArrayList<>();

    private final List<Link> relations = new ArrayList<>();
    private final List<Release> releases = new ArrayList<>();

    public List<Release> getReleases() {
        return releases;
    }

    public void setReleases(List<Release> releases) {
        this.releases.addAll(releases);
    }

    public List<Link> getRelations() {
        return relations;
    }

    public void setRelations(List<Link> relations) {
        this.relations.addAll(relations);
    }

    public String getTitle() {
        return title;
    }

    public void setTitle(String title) {
        this.title = title;
    }

    public String getPrimaryType() {
        return primaryType;
    }

    public void setPrimaryType(String primaryType) {
        this.primaryType = primaryType;
    }

    public int getCount() {
        return count;
    }

    public void setCount(int count) {
        this.count = count;
    }

    public List<ArtistCredit> getArtistCredits() {
        return artistCredits;
    }

    public void setArtistCredits(List<ArtistCredit> artistCredits) {
        this.artistCredits.addAll(artistCredits);
    }

    public List<String> getSecondaryTypes() {
        return secondaryTypes;
    }

    public void setSecondaryTypes(List<String> secondaryTypes) {
        this.secondaryTypes.addAll(secondaryTypes);
    }

    public String getFullType() {
        if (secondaryTypes.size() != 0) {
            String type = primaryType;
            for (String string : secondaryTypes) type = type.concat(" + ").concat(string);
            return type;
        } else return primaryType;
    }
}

File: app/src/main/java/org/metabrainz/android/model/mbentity/ReleaseGroup.kt
package org.metabrainz.android.model.mbentity

import com.google.gson.annotations.SerializedName
import org.metabrainz.android.data.sources.api.entities.ArtistCredit
import org.metabrainz.android.data.sources.api.entities.Link
import java.util.*

class ReleaseGroup : org.metabrainz.android.model.mbentity.MBEntity() {
    var title: String? = null

    @SerializedName("primary-type")
    var primaryType: String? = null

    @SerializedName("secondary-types")
    private val secondaryTypes: MutableList<String> = ArrayList()
    var count = 0

    //TODO: Implement correct wrapper JSON
    @SerializedName("artist-credit")
    val artistCredits: MutableList<ArtistCredit> = ArrayList()
    @JvmField
    val relations: MutableList<Link> = ArrayList()
    @JvmField
    val releases: MutableList<Release> = ArrayList()
    fun getReleases(): List<Release> {
        return releases
    }

    fun setReleases(releases: List<Release>?) {
        this.releases.addAll(releases!!)
    }

    fun getRelations(): List<Link> {
        return relations
    }

    fun setRelations(relations: List<Link>?) {
        this.relations.addAll(relations!!)
    }

    @JvmName("getArtistCredits1")
    fun getArtistCredits(): MutableList<ArtistCredit> {
        return artistCredits
    }

    fun setArtistCredits(artistCredits: List<ArtistCredit>?) {
        this.artistCredits.addAll(artistCredits!!)
    }

    fun getSecondaryTypes(): List<String> {
        return secondaryTypes
    }

    fun setSecondaryTypes(secondaryTypes: List<String>?) {
        this.secondaryTypes.addAll(secondaryTypes!!)
    }

    val fullType: String?
        get() = if (secondaryTypes.size != 0) {
            var type = primaryType
            for (string in secondaryTypes) type = "$type + $string"
            type
        } else primaryType

    override fun toString(): String {
        return "ReleaseGroup{" +
                "title='" + title + '\'' +
                ", primaryType='" + primaryType + '\'' +
                ", secondaryTypes=" + secondaryTypes +
                ", count=" + count +
                ", artistCredits=" + artistCredits +
                ", relations=" + relations +
                ", releases=" + releases +
                ", mbid='" + mbid + '\'' +
                ", disambiguation='" + disambiguation + '\'' +
                '}'
    }
}